### DISCUSSION TITLE
How does autodetect work exactly?

### DISCUSSION THREAD

**Discussion by IdrisQe** [PrismLauncher/PrismLauncher#3212] [Q&A] [answered]
Hi, so I've been really lazy about updating to 9.X, just finally did it, and I'm curious about how autodetect actually works.

So, on a fresh Windows (new Laptop) I installed Adoptiom JDK 8 and 17 (jdks becasue I may get back into making mods), then installed PrismLauncher. I copied over my instances and settings from my Desktop.

I enabled "Autodetect Java version" in the Settings>Java>General tab
There is also the "Java Runtime" section - does it matter what I put there when it should be autodetecting for each instance individually?

There is also the Settings>Java>Management tab, but it seems to be unable to detect my JDK installs. It says "No managed Java versions are installed". When I check the download options, it doesn't seem to have any JDKs available to download despite having a column that shows the type (JRE or JDK). Do my Java installs need to be detected in the Management page to work with Autodetection?

Should I keep the "Java Installation" checkbox unchecked on my instances' setting menus in order for Autodetection to work? I notice there's also the "Auto-detect.." button below that, is that related or is that just for manually selecting from an automatically generated list? (It seems to be that, just confused by the name)

Sorry, just a bit perturbed because there are like 3 different places that have options called some variant of "Auto-Detect" and it's difficult to parse what each does.

**Comment by Kationor** [comment, marked as answer]
> There is also the "Java Runtime" section - does it matter what I put there when it should be autodetecting for each instance individually?

What you put in there doesn't really matter.

> Do my Java installs need to be detected in the Management page to work with Autodetection?

No. Java management and autodetection work separately. Since 9.0, Prism can install and manage *its own* Java runtimes - those will show up there. Autodetection uses *both* runtimes managed by Prism *and* other runtimes that are installed on the system.

> Should I keep the "Java Installation" checkbox unchecked on my instances' setting menus in order for Autodetection to work?

Autodetection will automatically check it once it selects the right Java version, you don't really need to worry about it.

> I notice there's also the "Auto-detect.." button below that, is that related or is that just for manually selecting from an automatically generated list?

That button is there to help in manually selecting a Java runtime (instead of having to look for the path yourself). It is related to the "autodetect" setting in that they both use the same logic to search for valid Java installations (I think), but that's about it.

> Sorry, just a bit perturbed because there are like 3 different places that have options called some variant of "Auto-Detect" and it's difficult to parse what each does.

"Auto-Select" would probably be a more fitting name for the setting...

I'm only like 60% confident that I answered everything, so if I missed anything just lmk (or wait for somebody else to answer, I guess)

**Reply by IdrisQe** [reply]
Thank you very much for the extensive response! Yep that answers pretty much everything.
Sorry for not marking the answer sooner, been very preoccupied lately and haven't been on Github in a hot minute.

I guess the only other question I have (which I didn't really ask admittedly) is: Are there plans to be able to download JDKs? I notice that the manager has "JRE" listed in type for all downloads, there are no JDKs available to download, so the type column seems redundant.

**Reply by Kationor** [reply]
Judging by [this conversation on Discord](https://discord.com/channels/1031648380885147709/1031682669945294869/1219978673239625789), probably not.
To summarise that conversation: JDKs would only really be needed by mod devs who should have Java installed on their system anyway, so there is no need for Prism to support them. Since the support for JDKs is there and just disabled, it can always be enabled if it turns out to be needed for some reason.
